Dockerfile

```diff
@@ -59,30 +63,81 @@ RUN --mount=type=cache,target=/opt/ccache \
-         *)              pip install --index-url https://download.pytorch.org/${INSTALL_CHANNEL}/${CUDA_PATH#.}/ torch torchvision torchaudio ;; \
-    esac && \
+# Install necessary tools for dependency extraction
+RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+    jq \
+    && rm -rf /var/lib/apt/lists/*
+
+# Define packages to install
+ENV PACKAGES="torch torchvision torchaudio"
+
+# Step 1: Preinstall Torch Dependencies (Cached Separately)
+RUN set -eux; \
+    # Determine PIP_PRE_FLAG based on INSTALL_CHANNEL
+    if echo "${INSTALL_CHANNEL}" | grep -q "nightly"; then \
+    PIP_PRE_FLAG="--pre"; \
+    else \
+    PIP_PRE_FLAG=""; \
+    fi; \
+    # Determine EXTRA_INDEX_URL based on TARGETPLATFORM
+    if [ "$TARGETPLATFORM" = "linux/arm64" ]; then \
+    EXTRA_INDEX_URL="https://download.pytorch.org/whl/cpu/"; \
+    else \
+    EXTRA_INDEX_URL="https://download.pytorch.org/${INSTALL_CHANNEL}/${CUDA_PATH%.}/"; \
+    fi; \
+    echo "Using EXTRA_INDEX_URL=${EXTRA_INDEX_URL} with PIP_PRE_FLAG=${PIP_PRE_FLAG}"; \
+    # Create a temporary directory to store dependencies
+    TEMP_DIR=$(mktemp -d); \
+    trap 'rm -rf "$TEMP_DIR"' EXIT; \
+    # Initialize an empty requirements file for dependencies
+    DEP_REQUIREMENTS="${TEMP_DIR}/dependencies.txt"; \
+    touch "$DEP_REQUIREMENTS"; \
+    # Iterate over each package to collect dependencies
+    for PACKAGE in $PACKAGES; do \
+    echo "Fetching dependencies for $PACKAGE"; \
+    # Fetch package metadata from PyPI
+    curl -s https://pypi.org/pypi/${PACKAGE}/json -o "${TEMP_DIR}/${PACKAGE}.json"; \
+    \
+    # Extract requires_dist entries, preserving environment markers
+    jq -r '.info.requires_dist[]?' "${TEMP_DIR}/${PACKAGE}.json" >> "$DEP_REQUIREMENTS"; \
+    \
+    rm -f "${TEMP_DIR}/${PACKAGE}.json"; \
+    done; \
+    \
+    # Remove duplicate dependencies while preserving environment markers
+    sort -u "$DEP_REQUIREMENTS" -o "$DEP_REQUIREMENTS"; \
+    \
+    # Install dependencies using pip with the requirements file
+    if [ -s "$DEP_REQUIREMENTS" ]; then \
+    echo "Installing dependencies from $DEP_REQUIREMENTS"; \
+    pip install --no-cache-dir $PIP_PRE_FLAG --extra-index-url "$EXTRA_INDEX_URL" -r "$DEP_REQUIREMENTS"; \
+    else \
+    echo "No dependencies to install."; \
+    fi;
+# Step 2: Install Torch packages without dependencies
+RUN set -eux; \
+    if echo "${INSTALL_CHANNEL}" | grep -q "nightly"; then \
+    PIP_PRE_FLAG="--pre"; \
+    else \
+    PIP_PRE_FLAG=""; \
+    fi; \
+    if [ "$TARGETPLATFORM" = "linux/arm64" ]; then \
+    EXTRA_INDEX_URL="https://download.pytorch.org/whl/cpu/"; \
+    else \
+    EXTRA_INDEX_URL="https://download.pytorch.org/${INSTALL_CHANNEL}/${CUDA_PATH%.}/"; \
+    fi; \
+    echo "Installing main packages with PIP_PRE_FLAG=${PIP_PRE_FLAG}"; \
+    pip install --no-cache-dir $PACKAGES --no-deps $PIP_PRE_FLAG --extra-index-url $EXTRA_INDEX_URL && \
```

Comment:
Lets use ``--index-url`` here, if something is missing in the index I will upload to it.